modules/core/04-channel/keeper/upgrade.go

```diff
@@ -0,0 +1,79 @@
+package keeper
+
+import (
+	errorsmod "cosmossdk.io/errors"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	"github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v7/modules/core/24-host"
+)
+
+// ChanUpgradeInit is called by a module to initiate a channel upgrade handshake with
+// a module on another chain.
+func (k Keeper) ChanUpgradeInit(
+	ctx sdk.Context,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	proposedUpgradeChannel types.Channel,
+	counterpartyTimeoutHeight clienttypes.Height,
+	counterpartyTimeoutTimestamp uint64,
+) (uint64, error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
```

Comment:
Would it make sense to fail fast if the parameters of the current channel (ordering, version, connection hops) and the proposed new channel are exactly the same? Is it worth initiating the handshake if they are exactly the same? @AdityaSripal

Reply:
Good point! There may be a use case for updating one channel end's connection maybe? In that scenario we may want to allow one side to proceed even if they are identical.

Reply:
Yeah, I thought about that as well, but then would it still make sense that on init side the channel parameters are the same but on the try side they are different? Like would it work if the try side decides to change ordering unilaterally, for example? And for the version, for example try side upgrading to use fee middleware?

I see in the spec that on ack the ordering should be the same for both sides:

```
// both channel ends must be mutually compatible.
// this means that the ordering must be the same and 
// any future introduced fields that must be compatible
// should also be checked
if counterpartyChannel.ordering != proposedUpgradeChannel.ordering {
  restoreChannel(portIdentifier, channelIdentifier)
  return
}
```

Reply:
> Yeah, I thought about that as well, but then would it still make sense that on init side the channel parameters are the same as the current channel but on the try side they are different?

I agree with you that if only one side is changing (connection only I guess) then they are/should be the initiating party. Happy to add a `reflect.DeepEquals()` check here if there's consensus on that! 

> In the case of changing ordering, if the try side decides to change ordering unilaterally, I think the handshake will already fail on try.

Yep, you cannot change order unilaterally. Both channel ends must use the same ordering. 

> In the case of changing versions, if init side keeps the current channel version (let's say ics20-1 if we talk about transfer), but try side decides to use fee middleware

Yeah if you want to use ics29, both channel ends need to use it.

Reply:
ack'ing this comment thread, the additional check on connection hops looks good to me. 👍